src/db.c

```diff
@@ -851,21 +852,22 @@ void scanCallback(void *privdata, const dictEntry *de) {
     }
 
     if (o == NULL) {
-        listAddNodeTail(keys, createStringObject(keysds, sdslen(keysds)));
-        return;
+        key = keysds;
     } else if (o->type == OBJ_SET) {
-        /* Do nothing. */
+        key = keysds;
     } else if (o->type == OBJ_HASH) {
-        val = sdsdup(dictGetVal(de));
+        key = keysds;
+        val = dictGetVal(de);
     } else if (o->type == OBJ_ZSET) {
         char buf[MAX_LONG_DOUBLE_CHARS];
         int len = ld2string(buf, sizeof(buf), *(double *)dictGetVal(de), LD_STR_AUTO);
+        key = sdsdup(keysds);
```

Comment:
maybe instead of duping this one, we can have zset release only the values, and not the keys?
not sure if it's worth the trouble, have a look and decide.

Reply:
I could try it out but it feels a bit hard and code might get a bit ugly.

Reply:
For this remaining problem, I have a preliminary idea, hoping to add an addReplyBulkDouble to directly output double‘s strings, avoiding malloc/free,  and I hope the function can  be used in other places, not sure if it is feasible.